File: 18주차/이연상/올바른 괄호.py
# [PRG] 올바른 괄호

def solution(s):
    answer = True
    
    res = 0
    for i in s:
        if(i == "("):
            res += 1
        else:
            res -= 1
        
        if res < 0:
            return False
    
    if res > 0:
        return False

    return True

File: 39주차/이연상/올바른 괄호.java
// [PRG] 올바른 괄호

import java.util.*;

class Solution {
    boolean solution(String s) {
        boolean answer = true;

        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < s.length(); i++) {  
            if(s.charAt(i) == ')') {
                if(!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                    continue;
                }
            }
            stack.push(s.charAt(i));
        }
        if(!stack.isEmpty()) answer = false;
        
        return answer;
    }
}